### DISCUSSION TITLE
When I insert rows in the DataGrid, the order of the row numbers is wrong

### DISCUSSION THREAD

**Discussion by ApaulMyLittleAirport** [AvaloniaUI/Avalonia#9667] [Q&A] [answered]
When I insert rows in the DataGrid, the order of the row numbers is wrong, I want to ask how to set the row number of a given row. I can be in the WPF:
 "var gdr = (DataGridRow) MyDataGrid. ItemContainerGenerator. ContainerFromIndex (MyDataGrid. SelectedIndex)"
To get a Row and modify "row.header". But I can't find it in avalonia. Can anyone help me? Thank you very much!

**Comment by timunie** [comment, marked as answer]
Hi @ApaulMyLittleAirport maybe you can just create a Style for DataGridRow where you bind your item number (can also be a converter) to RowHeader?

**Reply by maxkatz6** [reply]
This also should work, setting up a binding in the styles.

**Reply by ApaulMyLittleAirport** [reply]
> Hi @ApaulMyLittleAirport maybe you can just create a Style for DataGridRow where you bind your item number (can also be a converter) to RowHeader?

Thank you very much for your reply, I'll give it a try

**Reply by Jakobinec** [reply]
Could you write an example please

**Reply by alvahtin** [reply]
```
<DataGrid>
     <DataGrid.Styles>
      	<Style Selector="DataGridRow">
              <Setter Property="Header" Value="{Binding Converter={StaticResource NumerRow}, Mode=OneWay, RelativeSource={RelativeSource Self}}" />
    	</Style>
    </DataGrid.Styles>
    <DataGrid.Columns>
			.........
    </DataGrid.Columns>
</DataGrid>
```

Converter:

```
public class NumerRow : IValueConverter
{
	public object Convert(object? value, Type TargetType, object? parameter, CultureInfo culture)
	{
		try
		{
			if (value is DataGridRow row)
			{
				return (row.GetIndex() + 1).ToString();
			}
			else
			{
				return "*";
			}
		}
		catch
		{
			return "*";
		}
	}

	public object ConvertBack(object? value, Type TargetType, object? parameter, CultureInfo culture)
	{
		throw new NotImplementedException();
	}
}
```

But it is wrong. See https://github.com/AvaloniaUI/Avalonia/discussions/9667#discussioncomment-4374526

**Comment by maxkatz6** [comment]
> "var gdr = (DataGridRow) MyDataGrid. ItemContainerGenerator. ContainerFromIndex (MyDataGrid. SelectedIndex)"
To get a Row and modify "row.header". But I can't find it in avalonia. Can anyone help me? Thank you very much!

You need to use LoadingRow event to handle this event. This way is times more reliable than having ContainerFromIndex API (which still is nice to have but won't solve your task properly). The problem is that rows are recycled, and you can handle recycling with LoadingRow and UnloadingRow events.

**Reply by ApaulMyLittleAirport** [reply]
xaml:
DataGridTextColumn Binding="{Binding Header, RelativeSource={RelativeSource AncestorType={x:Type DataGridRow}, Mode=FindAncestor}}"
xaml.cs
 this.MyDataGrid.LoadingRow += (sender, e) => {   e.Row.Header = e.Row.GetIndex() + 1; };

Excuse me, when I delete a row or insert a row, the line numbers of the other rows cannot be changed.

For example, before deleting rows.
![Snipaste_2022-12-12_09-43-58](https://user-images.githubusercontent.com/45100828/206946859-81e78847-5c1d-4c40-a515-fc2626dd9bef.jpg)

After I delete the first line.

![Snipaste_2022-12-12_09-41-45](https://user-images.githubusercontent.com/45100828/206946741-95fc0f79-3d1d-49d5-9f8f-9ebafbf32664.jpg)

**Reply by AmberHan** [reply]
你后来怎么解决的？

**Reply by timunie** [reply]
https://github.com/AvaloniaUI/Avalonia/discussions/9667#discussioncomment-4366289 should work

**Reply by alvahtin** [reply]
Its don't work in this case (( When a row is inserted, the number is assigned and does not change. But when deleting, need to change the numbering of all rows that are below the deleted one.

**Reply by ApaulMyLittleAirport** [reply]
后来我在observerCollection里创建了索引的属性，然后绑定了一下。。
